File: LeetCode/binary_tree_right_side_view.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> rightSideView(TreeNode* root) {
        if(!root)
            return vector<int> ();
        
        vector<int> res;
        queue<TreeNode*> q;
        
        q.push(root);
        while(!q.empty()) {
            int s = q.size();
            while(s > 0) {
                TreeNode* node = q.front();
                q.pop();
                
                if(s == 1)
                    res.push_back(node->val);
                
                if(node->left)
                    q.push(node->left);
                if(node->right)
                    q.push(node->right);
                
                s--;
            }
        }
        
        return res;
    }
};

File: LeetCode/binary_tree_right_side_view.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

type Queue struct {
    arr []*TreeNode
}

func (q *Queue) push(node *TreeNode) {
    q.arr = append(q.arr, node)
}

func (q *Queue) pop() *TreeNode {
    tmp := q.arr[0]
    q.arr = q.arr[1:]
    return tmp
}

func (q *Queue) size() int {
    return len(q.arr)
}

func rightSideView(root *TreeNode) []int {
    if root == nil {
        return []int{}
    }
    
    queue := &Queue{
        arr: []*TreeNode{root},
    }

    ans := []int{}
    for queue.size() > 0 {
        size := queue.size() 
        for i := 0; i < size; i++ {
            node := queue.pop()

            if i == size-1 {
                ans = append(ans, node.Val)
            } 

            if node.Left != nil {
                queue.push(node.Left)
            }

            if node.Right != nil {
                queue.push(node.Right)
            }
        }
    }

    return ans
}
